iseq.c

```diff
@@ -246,85 +240,22 @@ rb_vm_insn_normalizing_translator(const void *addr)
 typedef VALUE iseq_value_itr_t(void *ctx, VALUE obj);
 typedef VALUE rb_vm_insns_translator_t(const void *addr);
 
-static int
-iseq_extract_values(VALUE *code, size_t pos, iseq_value_itr_t * func, void *data, rb_vm_insns_translator_t * translator)
-{
-    VALUE insn = translator((void *)code[pos]);
-    int len = insn_len(insn);
-    int op_no;
-    const char *types = insn_op_types(insn);
-
-    for (op_no = 0; types[op_no]; op_no++) {
-	char type = types[op_no];
-	switch (type) {
-          case TS_CDHASH:
-          case TS_ISEQ:
-          case TS_VALUE:
-            {
-              VALUE op = code[pos + op_no + 1];
-              if (!SPECIAL_CONST_P(op)) {
-                  VALUE newop = func(data, op);
-                  if (newop != op) {
-                      code[pos + op_no + 1] = newop;
-                  }
-              }
-            }
-            break;
-          case TS_IC:
-            {
-                IC ic = (IC)code[pos + op_no + 1];
-                if (ic->entry) {
-                    VALUE nv = func(data, (VALUE)ic->entry);
-                    if ((VALUE)ic->entry != nv) {
-                        ic->entry = (void *)nv;
-                    }
-                }
-            }
-            break;
-          case TS_IVC:
-          case TS_ICVARC:
-            {
-                IVC ivc = (IVC)code[pos + op_no + 1];
-                if (ivc->entry) {
-                    if (RB_TYPE_P(ivc->entry->class_value, T_NONE)) {
-                        rb_bug("!! %u", ivc->entry->index);
-                    }
-                    VALUE nv = func(data, ivc->entry->class_value);
-                    if (ivc->entry->class_value != nv) {
-                        ivc->entry->class_value = nv;
-                    }
-                }
-            }
-            break;
-          case TS_ISE:
-            {
-              union iseq_inline_storage_entry *const is = (union iseq_inline_storage_entry *)code[pos + op_no + 1];
-              if (is->once.value) {
-                  VALUE nv = func(data, is->once.value);
-                  if (is->once.value != nv) {
-                      is->once.value = nv;
-                  }
-              }
-            }
-            break;
-          default:
-            break;
-	}
-    }
-
-    return len;
-}
-
 static inline void
-iseq_scan_bits(iseq_bits_t bits, VALUE *code, iseq_value_itr_t *func, void *data)
+iseq_scan_bits(unsigned int page, iseq_bits_t bits, VALUE *code, iseq_value_itr_t *func, void *data)
```

Comment:
Can you explain why this had to be reverted? I mean it's fine I doubt it makes any measurable performance difference, it just seemed cleaner to me to get this out of the function.

Reply:
I think removing that variable required that we [advance the code pointer on each iteration](https://github.com/ruby/ruby/pull/6062/files#diff-97b5a800b9fd7e139e5a18f23b67dbcc041262e5314606e46ccbcb9934d3a966L390).  We need to index in to the `original_iseq` list at the same time as the `code` list, but we don't always have an `original_iseq` list to advance.  This just seemed like the easiest way to index in to both lists (but only sometimes).

If you can think of a good way to index in to both lists and eliminate this variable, then I'm all for it! I just couldn't figure a better way.